### ISSUE TITLE
JVM crash when closing a jep instance associated to a terminated thread

### ISSUE THREAD

**Issue by matteozanon** [mrj0/jep#22]
I have a cache of jep instances which are bound to the container thread that created them.
If the originating thread is disposed by the web container and at a later time I try to close the corresponding jep instance I experience the following JVM crash:

     # A fatal error has been detected by the Java Runtime Environment:
     #
     #  SIGSEGV (0xb) at pc=0x00007fba779f1285, pid=25044, tid=140441049380608
     #
     # JRE version: Java(TM) SE Runtime Environment (7.0_80-b15) (build 1.7.0_80-b15)
     # Java VM: Java HotSpot(TM) 64-Bit Server VM (24.80-b11 mixed mode linux-amd64 compressed   oops)
     # Problematic frame:
     # C  [libjep.so+0x13285]  pyembed_thread_close+0xb5

**Comment by ndjensen** [comment]
Thanks for the error report.  I wrote a simple unit test to try and repeat this, I'm getting a different message for the crash, but that may just be due to my platform and configuration.  I never fully understood why <code>Jep.close()</code> does not call <code>isValidThread()</code>, that predates me.  I added the comment here: https://github.com/mrj0/jep/blob/dev_3.4/src/jep/Jep.java#L1088

With my unit test I've found that you can safely call <code>close()</code> from a different thread if the original thread that created the Jep instance/sub-interpreter is still alive.  This is probably why <code>finalize()</code> was overriden by @mrj0, to give an extra layer of ensuring memory is freed up.  However, if the original thread has died, then calling <code>close()</code> does crash the entire JVM, which is rather bad.

The easy but not ideal route is to make <code>close()</code> call <code>isValidThread()</code> so callers have to close from the original thread while it's still alive.  That makes it consistent with every other method on Jep, but a downside is that diverges from past behavior and I presume some would consider that a limitation.  I'm trying to think if there's a way we could intercept the thread termination/death and call <code>close()</code> at that time, before the thread dies.

**Comment by ndjensen** [comment]
Ok I dug into this way deeper and discussed it with @bsteffensmeier.  Could you please try making the following change?  It fixes my unit test but we'd like to know if it fixes your scenario.

In pyembed.c's pyembed_thread_close(), change the env line like so:
```c
//env = jepThread->env;
 env = pyembed_get_env();
```
It's kind of interesting that we're not using the JNIEnv* that is passed into jep.c.  If the old thread died, then the env attached to the jepThread object may point at who knows what and cause the SIGSEGV.

**Comment by ndjensen** [comment]
Related questions for @mrj0 that came about due to this investigation:
1. It appears <code>Jep.finalize()</code> no longer has value.  [When we make a Jep instance, the instance becomes caller in pyembed.c, which gets a new global reference](https://github.com/mrj0/jep/blob/dev_3.4/src/jep/pyembed.c#L287).  Therefore finalize() cannot be called before that global reference is deleted, which happens when close is called.  Do you agree with deleting the method finalize()?
2. In some areas of code, we use jepThread->env while in other areas we use pyembed_get_env().  The former is based on a global python dictionary of thread ids to JNIEnv*, while the latter uses JNI calls.  They presumably result in the same env 99% of the time.  For the most part it seems pyembed.c is using jepThread->env for the C methods on the jep module called from python.  Everywhere else is using pyembed_get_env().  Is one way faster or superior to another?  Or is some of it leftover code from an early release?

**Comment by matteozanon** [comment]
Changing like you suggested 
`env = pyembed_get_env();`
I don't get the JVM crash anymore but an exception that is much better.
I report the exception here:
>Exception ignored in: <module 'threading' from '/usr/lib/python3.4/threading.py'>
>Traceback (most recent call last):
>File "/usr/lib/python3.4/threading.py", line 1296, in _shutdown
> _main_thread._delete()
>File "/usr/lib/python3.4/threading.py", line 1019, in _delete
>  del _active[get_ident()]

Are you sure this is properly releasing resources?

**Comment by ndjensen** [comment]
Do you have more of the stacktrace?  What is the message with the exception?  Is this coming from the Jep.close() call?

**Comment by matteozanon** [comment]
I miss the last row 
>KeyError: 140442423424768

Unfortunately this is all I have.
Yes this happens when, after the change, I call jep.close() of a death thread jep instance.

**Comment by ndjensen** [comment]
I'm not sure that's coming from jep.  The KeyError looks like it might be looking for a thread id.  Your line of code that is:
```python
del _active[get_ident()]
```
You should maybe check
```python
id = get_ident()
if id in _active:
   del _active[id]
else:
   logger.warn("where did my thread " + str(id) + " go?")
```
Something like that.  Can you check that?  At the moment I'm not seeing how Jep could produce that error, though I might be missing something.

**Comment by matteozanon** [comment]
You are right, it's my code that generate the exception. It's my fault that I didn't check since I thought the error was still on Jep side.
With the fix to the C code that you suggested I don't experience the JVM crash anymore .
I move the discussion of my custom error to google group.

**Comment by bsteffensmeier** [comment]
With the change to env we can fix the crash and keep the java side of things happy but there still is some problems on the python side that I don't think we can fix.

1. Many of our threading problems in jep are related to the use of autoTLSkey in pystate.c in the cpython source. This key is used to track a single PyThreadState* for every running thread. For Jep.close() this causes problems when jep calls Py_EndInterpreter which calls _PyGILState_Fini() which calls PyThread_delete_key(autoTLSkey). The problem is that this deletes the PyThreadState* for the currently running thread, not for the thread that we are closing. This means we may have deleted a PyThreadState* that we needed and we are leaking a PyThreadState* that should be removed. This won't cause immediate problems because jep internally doesn't use the autoTLSkey but it causes parts of the python API to become unstable, for example the PyGILState_Ensure() cpython call can cause freezes, this has bitten us before in numpy when multiple Jep instances are opened on the same thread, causing in consistency in values for autoTLSkey.
2. It looks like the problem with threading is because the threading module is automatically shutdown when jep calls Py_EndInterpreter. threading makes the same assumption that the currently running thread is the one being shutdown and it tries to remove the currently running thread from the active threads. This has the same problems of potentially removing items we need while failing to remove the thread we are deleting, or in the case the closing thread has no jep instance you will get a key error.

Since closing the interpreter on a different thread causes inconsistent state I think it should be avoided. We should do something similar to what we did when initializing multiple interpreters on one thread and throw an exception or at least log a strong warning.

TL;DR Cpython doesn't support closing an interpreter on a different thread than the one that created it so I don't think jep can support close on a different thread.

**Comment by ndjensen** [comment]
Thanks @bsteffensmeier for the deeper analysis.  So that accounts for the stacktrace having issues with threading.py line 1296.  I don't see a safe way to solve this so I'm leaning towards making the following changes for Jep 3.4:

1. Update pyembed_thread_close() to use the JNIEnv* passed in instead of the JNIEnv* associated with the jep thread.  This fixes the crash.
2. Remove Jep.finalize() since it doesn't do anything until after close() is called and it complicates understanding.
3. Add a System.err.println() to Jep.close() if close() is called from a different thread.  This would be similar to [the warning in the Jep constructor](https://github.com/mrj0/jep/blob/dev_3.4/src/jep/Jep.java#L265) indicating that the call is unsafe from a separate thread, but would still allow it through for the time being.

**Comment by matteozanon** [comment]
Many thanks for the good explanation. With this premises I think I have to redefine how I manage my Jep instances.

**Comment by ndjensen** [comment]
I'm closing this as fixed in 3.4 since this scenario will not cause a crash anymore.

**Comment by SemanticBeeng** [comment]
Getting this from application test code.
Using `specs2` to run `jep` which calls a simple python script.

From sbt, first time it runs ok but second time it cashes with segmentation violation.
Any thoughts please @ndjensen ?

```
info] [sbt-avro4s] Found 0 schemas
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f40cae294e0, pid=14851, tid=0x00007f404a0dc700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libpython3.6m.so+0x19d4e0]  threadstate_getframe+0x0
#
# Core dump written. Default location: /development/projects/.../foundation/core or core.14851
#
# An error report file with more information is saved as:
# /development/projects/.../hs_err_pid14851.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
```

```
package com.comp.datalayer

import io.comp.datalayer.Code
import io.comp.jep.Scalathon
import jep.Jep
import org.slf4j.LoggerFactory
import org.specs2.execute.{AsResult, Result}
import org.specs2.mutable.Specification
import org.specs2.specification.ForEach

trait callContext extends ForEach[Jep] {


  def foreach[R: AsResult](f: Jep => R): Result = {

    implicit val jep: Jep = new Jep()
    try {
      println(s"spec start run ${jep} on thread ${Thread.currentThread().getName}")
      AsResult(f(jep))
    } finally {
      println(s"spec end end ${jep} on thread ${Thread.currentThread().getName}")
      jep.close()
    }
  }
}


class AnalysisOrchestrationSpec extends Specification with callContext {

  sequential

  protected val logger = LoggerFactory.getLogger("io.comp.orchestration")

  "orchestration spec" should {

    "extract some feature" >> { jep: Jep =>

      runExtractFeature1(jep)
      ok
    }

  }

  private def runExtractFeature1(implicit jep: Jep): Unit = {

    import Scalathon._

    pythonImport(s"${Code.path}/extract_features")
    val feature1: Long = 'extract_feature1 (Seq(100))
    println(feature1)

  }
}
```

```
from java.lang import Integer
from java.util import ArrayList as AL

def extract_feature1(param1):

    print(f">>>>>>> Extracted feature {param1}")

    x = Integer(5)
    y = Integer(51)
    a = AL()

    return 1
```
